---
title: "GenesRanking_vignette"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{GenesRanking_vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```
# GenesRanking

  - [GenesRanking](#GenesRanking)
  - [Introduction](#Introduction)
  - [Installation](#Installation)
  - [Usage-example](#Usage-example)


### Introduction
The GenesRanking package offers advanced algorithms for feature selection
in high-dimensional biological data. These include statistical approaches
and various machine learning methods using tree-based or non-tree-based
approaches. Additionally, the package provides functions for data
normalization, pre-processing, and enrichment analysis using the
clusterProfiler package.

Our package offers a comprehensive set of tools to facilitate users' first
experience with single-cell RNA sequencing data analysis through feature
selection techniques. The ExampleDataset, Labels, and FilterData objects
are provided to assist users. Labels are character or factor vectors that
identify each cell in the data matrix, while ExampleDataset is a data.frame
object capturing gene expression data for individual cells. FilterData is
a list object containing filtered and/or processed gene expression data.

To analyze the data, users may choose from two key approaches:
machine learning and statistical methods. However, before applying
any analysis techniques, data preprocessing steps such as normalization 
and/or noise reduction may be required to improve data quality, 
reliability, and to reduce computational time.


### Installation
```
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install(GenesRanking)

```

```{r sessionInfo, warning=FALSE, message=FALSE,results='hide'}
library(GenesRanking)
sessionInfo()
```

### Usage-example

To demonstrate the use of our package, we first loaded the necessary libraries
and example dataset.
```{r libraries, warning=FALSE, message=FALSE}
library(GenesRanking)
data("ExampleDataset")
data("Labels")
data("FilterData")
```

Next, we proceeded to preprocess our data and demonstrate our approach for each
category.
```{r Preprocess, echo=T, message=F, warning=F}
# Apply normalization methods such as log2 normalization.
Norm_data=normalize_data(ExpressionData=ExampleDataset, 
                         normal_method="LogNormalize") 

# Utilizing a high variability filter serves to increase the reliability of our
# data by reducing the impact of noisy or irrelevant features.
data_filtered <- filter_genes(data=ExampleDataset, 
                 filter_method = "High_Variable_Genes", n_features = 2000)
```

We have at our disposal three distinct methods for identifying genes with
dominant expression patterns. For illustrative purposes, we will provide an
example demonstrating each method.

```{r Statistical-Non_Tree_Based_ML-Tree_Based_ML, echo=T, message=F, warning=F}
#Statistical_Filter 
StatisticalFilter <- Statistical_Filter(data_filtered, Pvalue_md = "Waldtest",
                     Labels = Labels, n_genes_to_keep = 100)

#LD_Based_ML
result_LD_Based_ML <- LD_Based_ML(data_filtered, Labels=Labels,
                                              MLmethod = "knn",  n_genes_to_keep = 100)

#Tree_Based_ML
result_Tree_Based_ML <- Tree_Based_ML(data_filtered, Labels=Labels, 
                        MLmethod = "C5.0", n_genes_to_keep = 100)

#Boosting_Based_ML
result_XGBoost_Based_ML <-  Boosting_Based_ML(data_filtered,Labels=Labels, 
                                     "xgbTree",n_genes_to_keep = 100)


```


Let's examine these lists of dominant genes, which can be used for further downstream analysis, such as functional annotation or pathway enrichment analysis.
```{r Dominants genes}
# Dominant genes identified by each method:
# Each of the methods has identified 100 dominant genes, providing insights into the most relevant features in the dataset.

# Display the top 10 dominant genes found using the Statistical_Filter method
head(StatisticalFilter$Important_Features, 10)

# Display the top 10 dominant genes found using the LD_Based_ML method
head(result_LD_Based_ML$Important_Features, 10)

# Display the top 10 dominant genes found using the Tree_Based_ML method
head(result_Tree_Based_ML$Important_Features, 10)

# Display the top 10 dominant genes found using the Boosting_Based_ML method
head(result_XGBoost_Based_ML$Important_Features, 10)

```

Now, we can perform an enrichment analysis using the GO ontology on the identified dominant genes from the previous methods. Let's take a look at an example.
```{r Ontology_Analysis genes, echo=T, message=F, warning=F}
OntologyAnalysis= Ontology_Analysis(
  Important_Features = result_XGBoost_Based_ML$Important_Features,
                       ontology_type = "GO",
                       organism = "mouse")

head(OntologyAnalysis)


```
