---
title: "GenesRanking_vignette"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{GenesRanking_vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```
# GenesRanking

  - [GenesRanking](#GenesRanking)
  - [Introduction](#Introduction)
  - [Installation](#Installation)
  - [Usage-example](#Usage-example)


### Introduction

The GenesRanking package offers a range of advanced algorithms for feature
selection in high-dimensional biological data. These include statistical
approaches, established methods for identifying variable genes, and a variety
of machine learning methods using wrapper or tree-based approaches.
Additionally, the package provides functions for data normalization and
pre-processing to ensure reliable data analysis, as well as enrichment analysis
using the clusterProfiler package.

Our package offers a comprehensive set of tools to facilitate the user's first
experience with the single-cell RNA sequencing data analysis. Specifically,
ExampleDataset, Labels, and FilterData are provided to assist the
user to have his first experience with our package. Labels are character or
factor vectors that identify each cell in the data matrix, while ExampleDataset
is a data.frame object that captures the gene expression data for individual
cells. FilterData, on the other hand, is a list object that contains filtered
and/or processed gene expression data.

To analyze the data, the user may choose from a range of four key approaches,
including variable gene methods, machine learning, and statistical methods.
However, before applying any analysis techniques, data preprocessing steps such
as normalization and/or noise reduction may be required to improve data quality
and reliability.

### Installation
```{r sessionInfo, warning=FALSE, message=FALSE}
library(GenesRanking)
sessionInfo()
```

### Usage-example

To demonstrate the use of our package, we first loaded the necessary libraries
and example dataset.
```{r libraries, warning=FALSE, message=FALSE}
library(GenesRanking)
data("ExampleDataset")
data("Labels")
data("FilterData")
```

Next, we proceeded to preprocess our data and demonstrate our approach for each
category.
```{r Preprocess, echo=T, message=F, warning=F}
# Apply normalization methods such as log2 normalization.
Norm_data=normalize_data(ExpressionData=ExampleDataset, 
                         normal_method="LogNormalize") 

# Utilizing a high variability filter serves to increase the reliability of our
# data by reducing the impact of noisy or irrelevant features.
data_filtered <- filter_genes(data=ExampleDataset, 
                 filter_method = "High_Variable_Genes", n_features = 1000)
```
We have at our disposal four distinct methods for identifying genes with
dominant expression patterns. For illustrative purposes, we will provide an
example demonstrating each method.

```{r Statistical-Non_Tree_Based_ML-Tree_Based_ML, echo=T, message=F, warning=F}
#Statistical_Filter 
StatisticalFilter <- Statistical_Filter(data_filtered, Pvalue_md = "Waldtest",
                     Labels = Labels, n_genes_to_keep = 100)

#Non_Tree_Based_ML
result_Non_Tree_Based_ML <- Non_Tree_Based_ML(data_filtered, Labels=Labels,
                                              MLmethod = "knn",  n_genes_to_keep = 100)

#Tree_Based_ML
result_Tree_Based_ML <- Tree_Based_ML(data_filtered, Labels=Labels, 
                        MLmethod = "C5.0", n_genes_to_keep = 100)



```

```{r Dominants genes}
#Dominants genes isolated eache method:
#Each of the methods has identified 100 dominant genes, with the exception of the variable genes method.

head(StatisticalFilter$Important_Features,10)

head(result_Non_Tree_Based_ML$Important_Features,10)

head(result_Tree_Based_ML$Important_Features,10)

```

Now, we can perform an enrichment analysis on the identified dominant genes.
Let's take a look at an example.
```{r Ontology_Analysis genes, echo=T, message=F, warning=F}
OntologyAnalysis= Ontology_Analysis(
  Important_Features = FilterData$Important_Features[1:10],
                       ontology_type = "GO",
                       organism = "mouse")

head(OntologyAnalysis)


```
